README
DestructAssign version 0.01
===========================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2014 by Cindy Wang (CindyLinz)

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension DestructAssign.

0.2.6    2016.8.27
      Fix bug by version 0.2.5

0.2.5    2016.8.26
      Compatible with newer (5.25 up) perl

0.2.4    2015.5.25
      Move '-std=gnu99' from CCFLAGS to DEFINE in Makefile.PL to avoid Perl 5.22 XS module handshake problem.

0.2.3    2014.12.10
      Compatible with Perl 5.21.7.
      Patched by Father Chrysostomos (sprout@cpan.org)

0.2.2    2014.12.8
      Compatible with the op tree of Perl 5.21.6.
      More informative warning message when matching against in-compatible value type.

0.2.1    2014.11.17
      Fix bug, which occurs when using variable name as key
      at the first element of a hash pattern

0.2      2014.11.3
      Bump version

0.1.4    2014.11.3
      Now we can use hash pattern to match against a array reference
      Use names of variables in hash patterns as keys when not specified keys for them.
      Use previously used key when not specified a key for a sub-pattern in a hash pattern.

0.1.3    2014.10.20
      Fix bug when use pattern in a recursive sub

0.1.2    2014.8.21
      Fix bug when the first element of an anonlist or an anonhash pattern
      is a pattern.

0.1.1    2014.8.12
      Fix document

0.1  Tue Aug  5 15:11:47 2014
	- original version; created by h2xs 1.23 with options
		-n DestructAssign



try.pl
use strict;
use warnings;

use B::Deparse;

use DestructAssign qw(des des_alias);
use Carp;
use Data::Dumper;
$Data::Dumper::Indent = 0;
$Carp::MaxArgLen = 0;
$Carp::MaxArgNums = 0;

BEGIN {
    $SIG{__DIE__} = sub { confess $_[0] };
}

our $y;

sub tt($) : lvalue {}

#{
#    my($x, $z, $q, $g, @remain, %hv, $o);
#    tt [[$x, 2 => $y, { a => $z, b => $A::o }, @A::X, %B::Y, \%hv, \$y, \$x, [$x]], {a => $q}];
#}

# 預期用法
# {
#    my($a, $b, $c);
#    des [$a, {b => $b, c => [undef, $c]] = [1, {a => 'a', b => [1,2,3], c => [5, 6]];
#    # 得到
#    # $a = 1
#    # $b = [1,2,3]
#    # $c = 6
# }
# array 加上 jump index
# {
#    my($a, $b, $c, $d);
#    des [$a, 2, $b, $c, 9, $d] = [0, 1, 2, 3, 4]
#    # 得到
#    # $a = 0
#    # $b = 2
#    # $c = 3
#    # $d = undef
# }
# 順便幫忙 my
# {
#    des_my [$a, {b => $b, c => [undef, $c]] = [1, {a => 'a', b => [1,2,3], c => [5, 6]];
#    # 得到 (而且都自動加上 lexical 宣告)
#    # $a = 1
#    # $b = [1,2,3]
#    # $c = 6
# }
# alias 型式
# {
#    my $data = [1, {a => 'a', b => [1,2,3], c => [5, 6]];
#
#    des_my_alias [$a, {b => $b, c => [undef, $c]] = $data;
#    # or
#    my ($a, $b, $c);
#    des_alias [$a, {b => $b, c => [undef, $c]] = $data;
#
#    $a = 2; $b = { b => 'b' }; $c = 'oo';
#    # 得到
#    # $data = [2, {a => 'a', b => {b => 'b'}, c => [5, 'oo']];
# }

sub f {
    my($q, $g, @remain, %hv, $o);
    des [my $x, $y, my $z] = \@_;
    print "\$x=$x \$y=$y \$z=$z\n";
    des [undef, $x, -2 => $z, [$y]] = [3, 4, 5, [6], 8, [7]];
    print "\$x=$x \$y=$y \$z=$z\n";
    des {x => $x, y => $y} = {x => 'x', y => 'y'};
    print "\$x=$x \$y=$y \$z=$z\n";
    @remain = ('a'..'c');
    des [$x, {oo => [$A::o]}, -5 => @remain, -5 => %hv, -5 => @A::array, -4 => %A::hash] = ['x', {oo => ['Ao']}, 1..9];
    print "\$x=$x \$A::o=$A::o \@remain=@remain \@A::array=@A::array\n";
    local $Data::Dumper::Indent = 0;
    print '%hv=', Dumper(\%hv), $/;
    print '%A::hash=', Dumper(\%A::hash), $/;
    #des [ *main::yy ] = [sub { print "yy\n" }];

    #yy();
}

BEGIN {
    my $depth = 0;
    my $silent = 0;
    for my $entry ( keys %B::Deparse:: ) {
        no strict 'refs';
        if( my $f = *{"B::Deparse\::$entry"}{CODE} ) {
            no warnings 'redefine';
            no warnings 'prototype';
            *{"B::Deparse\::$entry"} = sub {
                if( $silent ) {
                    return &$f;
                }
                if( grep { $entry eq $_ } qw(pessimise class null is_state is_scope is_miniwhile is_for_loop) ) {
                    my $ori_silent = $silent;
                    $silent = 1;
                    if( wantarray ) {
                        my @res = &$f;
                        $silent = $ori_silent;
                        return @res;
                    } elsif( defined wantarray ) {
                        my $res = &$f;
                        $silent = $ori_silent;
                        return $res;
                    } else {
                        &$f;
                        $silent = $ori_silent;
                        return;
                    }
                }
                my($package, $filename, $line) = caller;
                for (1..$depth) {
                    print " ";
                }
                local $" = ',';
                print "> $entry(@_) from $filename:$line\n";
                ++$depth;
                my $leave = sub {
                    --$depth;
                    for (1..$depth) {
                        print " ";
                    }
                    if( defined $_[0] ) {
                        print "< $entry -> ($_[0])\n";
                    } else {
                        print "< $entry\n";
                    }
                };

                if( wantarray ) {
                    my @res = &$f;
                    if( $entry eq 'gv_name' || $entry eq 'padname' || $entry eq 'const' ) {
                        $leave->($res[0]);
                    } else {
                        $leave->();
                    }
                    return @res;
                } elsif( defined wantarray ) {
                    my $res = &$f;
                    if( $entry eq 'padname_sv' || $entry eq 'gv_name' || $entry eq 'padname' || $entry eq 'const' ) {
                        $leave->($res);
                    } else {
                        $leave->();
                    }
                    return $res;
                } else {
                    &$f;
                    $leave->();
                    return;
                }
            };
        }
    }

    #print B::Deparse->new('-p', '-P')->coderef2text(\&f),$/;
}

f(qw(X Y Z));

{
    my $data = [1, {x => [2,3,4], y => 6}, 3];
    des_alias [my $a, {y => my $b, x => [2 => my $c]}] = $data;
    print '$data=', Dumper($data), $/;
    print "\$a=$a, \$b=$b, \$c=$c\n";
    ($a, $b, $c) = (4, 5, 6);
    print '$data=', Dumper($data), $/;
}

{
    des [my($a, $b), {a => my $c}] = [1,2,{a => 3}];
    print "$a $b $c\n";
}

{
    my $data = [1,2];
    my $a = 5;
    print "$a $data->[0]\n";
    {
        des_alias [my $a] = $data;
        print "$a $data->[0]\n";
        $a =20;
        print "$a $data->[0]\n";
    }
    print "$a $data->[0]\n";
    $a = 6;
    print "$a $data->[0]\n";
}

{
    my $a;
    des [[$a]] = [[1]];
    print "$a\n";
}

{
    my %hash = (a => 1, b => 2);
    my @array = ('a','b','c');

    use constant {
        A => 'a',
        B => 'b',
        ONE => 1,
        TWO => 2,
    };
    des{ A, my $a, B, my $b } = \%hash;
    des[ ONE, my $c, TWO, my $d ] = \@array;
    print "$a,$b,$c,$d\n";
}
use constant {
    O => 5,
};

{
    my $f = sub {
        my($acceptor) = @_;
        #$acceptor->{a} = 123;
        $_[0]{a} = 123;
    };
    $f->(des { a => my $a, b => my $b });
    print "a=$a, b=$b\n";
}

sub g {
    my $h = des { a => my $a, b => my $b } = { a => 1, b => 2 };
    print Dumper($h),$/;

    if( $_[0]>0 ) {
        g($_[0]-1);
    }
}

g(3);

for(1,2) {
    our $A;
    des { $A::b, $A, my $c } = { A => 2, b => 3, c => 4 };
    print $A,$/;
    print $A::b,$/;
    print $c,$/;
}

sub modify {
    des_alias[my($a, $b, $c)] = \@_;
    $c = $a + $b;
}

{
    my($a, $b, $c) = (1, 2, 0);
    modify($a, $b, $c);
    print $c, $/;
}

{
    my($a, $b, $c, $d, $x, $y, $z);
    des{
        $a, $b, $c,
        d => {
            $x, $y, $z
        },
    } = {a => 1, b => 2, c => 3, d => {x => 10, y => 11, z => 12}};
    print "$a $b $c $x $y $z\n";
}


Makefile.PL
use 5.008;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'DestructAssign',
    VERSION_FROM      => 'lib/DestructAssign.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/DestructAssign.pm', # retrieve abstract from module
       AUTHOR         => 'Cindy Wang (CindyLinz) <cindy@cpan.org>') : ()),
    LIBS              => [''], # e.g., '-lm'
    DEFINE            => '-std=gnu99', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I.', # e.g., '-I. -I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);
if  (eval {require ExtUtils::Constant; 1}) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (qw());
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'DestructAssign',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

}
else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}


lib/DestructAssign.pm
package DestructAssign;

use 5.008;
use strict;
use warnings;
use Carp;

require Exporter;
use AutoLoader;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use DestructAssign ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
    des des_alias
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.002006';

sub des($) : lvalue { $_[0] }
sub des_alias($) : lvalue { $_[0] }

require XSLoader;
XSLoader::load('DestructAssign', $VERSION);

# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__

=head1 NAME

DestructAssign - Destructuring assignment

=head1 SYNOPSIS

  use DestructAssign qw(des des_alias);

  my($w, $x, $y, $z);
  our($X, $Y, $Z);
  des [$x, [undef, {y => $y}, undef, $w], $z] = [2, [25, {x => 'x', y => 3}, 26, 1], 4];
  # got ($w, $x, $y, $z) = (1, 2, 3, 4)
  # (use undef as the skipping placeholder)

  # put skip index in the list pattern
  des [3 => $w, $x, -2 => $y, $z] = [1..9];
  # got ($w, $x, $y, $z) = (4, 5, 8, 9)

  # use names of the variables in a hash pattern as keys when not assigned
  # use previously used key for a sub-pattern when not assigned
  des {$x, $A::B, $Y, [$a, $b]} = {x => 1, Y => [9, 8], B => 3};
  # got ($x, $Y, $A::B, $a, $b) = (1, [9,8], 3, 9, 8);

  # use hash pattern to match against an array reference
  # So we can write:
  sub f {
    des {my($score, $name, $detail), {my($math, $english)}} = \@_;
    ...
  }
  f(
    name => 'Cindy',
    score => 95,
    detail => {math => 90, english => 30, bios => 60},
  );


  # put @array or @hash in the list pattern to eat all the remaining element
  my(@array, %hash);
  des [3 => @array, -4 => %hash] = [1..8];
  # got @array = (4..8), %hash = (5 => 6, 7 => 8)

  # put the same index or hash key
  #  when you need to capture different granularity on the same data structure
  #  (notice that you can use duplicated keys in the hash pattern)
  des {x => $x, x => [$y, $z]} = {x => [1, 2]};
  # got $x = [1,2], $y = 1, $z = 2

  # use the alias semantics
  my $data = [1, 2, 3];
  des_alias [undef, $x] = $data;
  $x = 20;
  # got $data = [1, 20, 3]

  {
    # mixed with lexical variable introduction
    des [my($i, $j), { k => my $k }] = [1, 2, {k => 3}];
    # got my($i, $j, $k) = (1, 2, 3)
  }

=head1 DESCRIPTION

This mod provides destructuring assignment for Perl.
You can capture (by value) or bind (by alias) variables into
part of a potentially large and complex data structure.

I expect it to bring following benefits:

=over 4

=item provide named parameters more easily

Named parameters are good when the number of parameters is large (more than 4).
With this mod, you can do:

  sub f {
    des {my($id, $title, $x, $y, $width, $height)} = \@_;
    # The order is not important.
    ...
  }

  f(
    id => 1,
    title => 'Untitled',
    x => 10, y => 10,
    width => 200, height => 150,
  );

=item enhance the readability by pointing out all the elements you might touch at the begining of each subroutine

It's a good habit to name parameters instead of access @_ directly
(except you want to modify caller's arguments).
This mod extend the ability to name parameters in the deep structure.
You can explicitly list all the elements you might touch in the subroutine.

  sub f {
    des [my $x, { id => my $id, amount => my $amount }] = \@_;
    # or use des_alias, if you need to modify the passed parameters.
    des_alias [my $x, { id => my $id, amount => my $amount }] = \@_;
  }

Even if you want to modify caller's arguments, you can still use "des_alias" to name them.

  sub add {
    des_alias [my($a, $b, $sum)] = \@_;
    $sum = $a + $b;
  }

  my($a, $b, $c) = (1, 2, 0);
  add($a, $b, $c);
  # $c = 3

=item enhance the performance by avoiding repeatedly digging into complex data structures

Suppose we have data structures like this:

  my $player1 = {
    id => 25,
    hp => 8100,
    armor => {
      body => {
        id => 21,
        name => 'iron suit',
        protect => 10,
        durability => 100,
      },
      hand => {
        id => 29,
        name => 'iron sword',
        attack => 15,
        durability => 100,
      },
    },
  };
  my $player2 = ...;

Instead of

  while( $player1->{hp}>0 && $player2->{hp}>0 ) {
    my $hit1 =
        ($player1->{armor}{hand}{durability} && $player1->{armor}{hand}{attack}) -
        ($player2->{armor}{body}{durability} && $player2->{armor}{body}{protect});
    my $hit2 =
        ($player2->{armor}{hand}{durability} && $player2->{armor}{hand}{attack}) -
        ($player1->{armor}{body}{durability} && $player1->{armor}{body}{protect});
    $hit1 = 1 if( $hit1 <= 0 );
    $hit2 = 1 if( $hit2 <= 0 );

    $player1->{hp} -= $hit2;
    $player2->{hp} -= $hit1;

    --$player1->{armor}{hand}{durability} if( $player1->{armor}{hand}{durability} );
    --$player1->{armor}{body}{durability} if( $player1->{armor}{body}{durability} );
    --$player2->{armor}{hand}{durability} if( $player2->{armor}{hand}{durability} );
    --$player2->{armor}{body}{durability} if( $player2->{armor}{body}{durability} );
  }

We could write

  des_alias [
    {
      hp => my $hp1,
      armor => {
        body => {
          protect => my $protect1,
          durability => my $body_dura1,
        },
        hand => {
          attack => my $attack1,
          durability => my $hand_dura1,
        },
      }
    },
    {
      hp => my $hp2,
      armor => {
        body => {
          protect => my $protect2,
          durability => my $body_dura2,
        },
        hand => {
          attack => my $attack2,
          durability => my $hand_dura2,
        },
      }
    },
  ] = [$player1, $player2];

  while( hp1>0 && hp2>0 ) {
    my $hit1 = ($hand_dura1 && $attack1) - ($body_dura2 && $protect2);
    my $hit2 = ($hand_dura2 && $attack2) - ($body_dura1 && $protect1);
    $hit1 = 1 if( $hit1 <= 0 );
    $hit2 = 1 if( $hit2 <= 0 );

    $hp1 -= $hit2;
    $hp2 -= $hit1;

    --$hand_dura1 if( $hand_dura1 );
    --$body_dura1 if( $body_dura1 );
    --$hand_dura2 if( $hand_dura2 );
    --$body_dura2 if( $body_dura2 );
  }

=back

This mod is tested in Perl 5.8.9, 5.10.1, 5.12.5, 5.14.4, 5.16.3, 5.18.2, 5.20.0, 5.21.6, 5.21.7 on x86_64.

=head2 EXPORT

None by default.

=over 4

=item des pattern = $value

The value semantics of destructuring assignment.

The captured data elements are copied out to the variables.

=item des_alias pattern = $value

The alias semantics of destructuring assignment

The captured data elements are bound to the variables.
After the binding, write to or read from the variables will
affect the bound data.

Be careful that once you bind a variable to a data element,
there's no easy way to unbind it.
It's recommended to use brand new lexical variables or localized variables to do it.

Like this..

  # some variables outside..
  my $a = 123;
  our $y = 456;

  my $data = [1, 2]; # the target data
  {
      des_alias [my $a, local $y] = $data;
      $a = 5; $y = 6;
      # $data = [5, 6];
  }
  # back to original, unbound $a and $y.
  $a = 7;
  $y = 8;
  # $data = [5, 6]; # the $data will not be changed.

=back

=head2 PATTERN

The only argument to C<des> or C<des_alias> should be
an anonymous list [..] or anonymous hash {..}.

The elements of them could be

=over 4

=item anonymous list [..]

To match a list reference

=item anonymous hash {..}

To match a hash reference or an array reference

=item scalar variable $xx

To capture the data element

=item array variable @xx

This can be used only in an anonymous list, not hash.

To capture all the remaining data element of the current array
from current offset.

When using in alias semantics, each element of the array variable will
be bound to data element individually.

=item hash variable %xx

This can be used only in an anonymous list, not hash.

To capture all the remaining data element of the current array
from current offset.

When using in alias semantics, only the values of the hash variable
will be bound to data element, individually.

=item constant undef

When using in an anonymous list, it is used as a skipping placeholder.

When using in an anonymous hash, it is used to set the next capturing key to ''.

=item constant string or number

When using in an anonymous list, we'll take it as a numer to set
the next capturing offset.

When using in an anonymous hash, we'll take it as a string to set
the next capturing key.

=back

=head1 SEE ALSO

This mod's github L<https://github.com/CindyLinz/Perl-DestructAssign>.
It's welcome to discuss with me when you encounter bugs, or
if you think that some patterns are also useful but the mod didn't provide them yet.

I also found a similar mod on github. (no cpan page) L<https://github.com/hirokidaichi/p5-Data-Destructuring-Assignment>
It's implemented in pure Perl with tied structures.
Because it's pure Perl, we need to pass references directly on the left hand side.
It can't take advantage on hash pattern with duplicated keys.
It didn't provide alias semantics either, though it could be added easily if needed.

=head1 AUTHOR

Cindy Wang (CindyLinz) <cindy@cpan.org>

=head1 CONTRIBUTOR

Father Chrysostomos <sprout@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014 by Cindy Wang (CindyLinz)

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.


=cut
